### PR TITLE
Fix startup with interface without IP

### PR DESCRIPTION
**Pull request by acksys**
If an interface get the IP information from a DHCP server, and we start
pimd before the interface have IP it will be never possible to forward
multicast paquet on this interface.

If phyint is set in configuration file, we loop while all interfaces
present in phyint have an IP address.
In general the IP will be received very soon.

Signed-off-by: Cedric Voncken <cedric.voncken@acksys.fr>

### CODE REVIEW

**Comment by troglobit** [comment]
Good catch!

I only have a few comments which would be great if you could address before I merge this PR.

File: src/config.c [hunk 1]
```diff
@@ -225,21 +228,70 @@ void config_vifs_from_kernel(void)
...
+init_vif_list:
     total_interfaces = 0; /* The total number of physical interfaces */
     if (getifaddrs(&ifaddr) == -1) {
 	logit(LOG_ERR, errno, "Failed retrieving interface addresses");
 	return;
     }
 
+    ifap = calloc(phyint_num, (sizeof(struct ifaddrs)));
+    if(!ifap) {
+	    logit(LOG_ERR,0,"%s[%d]: Allocation error",__func__,__LINE__);
```
Comment (troglobit) [inline src/config.c]: I'm a bit picky about following the coding style pf surrounding code:

1. `if(...)` --> `if (...)`
2. All arguments should be spaced, like in a regular sentence: `logint(LOG_ERR, 0, "...", __func__, __LINE__);`

File: src/config.c [hunk 2]
```diff
@@ -225,21 +228,70 @@ void config_vifs_from_kernel(void)
...
+	    return;
+    }
+
+    for (count=0,ifa = ifaddr; ifa; ifa = ifa->ifa_next) {
+	LIST_FOREACH(entry, &il, link) {
+//		logit(LOG_INFO,0,"name %s flag %x ifa_addr %p",ifa->ifa_name,ifa->ifa_flags,ifa->ifa_addr);
```
Comment (troglobit) [inline src/config.c]: Maybe a useful developer debug comment, but not needed in production code.  Remove or change to `LOG_DEBUG` so the default log level filters it out.

File: src/config.c [hunk 3]
```diff
@@ -225,21 +228,70 @@ void config_vifs_from_kernel(void)
...
+
+	    if(((ifa->ifa_flags & IFF_UP) != IFF_UP) || strcmp(entry->ifname,ifa->ifa_name))
```
Comment (troglobit) [inline src/config.c]: Again, lots of missing spaces.  Inconsistent also with your own style in the following if() statement.

File: src/config.c [hunk 4]
```diff
@@ -225,21 +228,70 @@ void config_vifs_from_kernel(void)
...
+				ifap[count-1].ifa_next = &ifap[count];
+			count++;
+			if (count > phyint_num){
+				logit(LOG_ERR,0,"find more interface than configured");
+			}
+	    }
+	}
+    }
+
+#if 0
```
Comment (troglobit) [inline src/config.c]: Another disabled developer debug message, remove altogheter or replace `#if 0`  with, e.g. `IF_DEBUG(l)` from `debug.h`

File: src/config.c [hunk 5]
```diff
@@ -225,21 +228,70 @@ void config_vifs_from_kernel(void)
...
+		addr  = ((struct sockaddr_in *)ifap[i].ifa_addr)->sin_addr.s_addr;
+		logit(LOG_INFO,0,"%s: IP %s ifap[%d]: addr %p next %p",ifap[i].ifa_name,
+				inet_fmt(addr, s1, sizeof(s1)),
+				i,&ifap[i],ifap[i].ifa_next);
+    }
+    logit(LOG_INFO,0,"------------------ %s end Ack dbg ----------",__func__);
+#endif
+
+    freeifaddrs(ifaddr);
+    if(!do_vifs && count < phyint_num){
```
Comment (troglobit) [inline src/config.c]: Another whitespace mishap.

Other than these comments, it looks good! :+1:

**Comment by acksys** [comment]
Thanks, I will fix my commit

**Comment by troglobit** [comment]
Some minor stuff remains, but I can fix that up myself later. Thanks! :)

For future reference, it's OK to push another commit instead of editing/rebasing the existing one.  This also makes it easier for the maintainer to review fixes to each comment made in the review.

**Comment by troglobit** [comment]
Hmm ... I just discovered that this change completely breaks the default use-case, running plain `pimd` without any `phyint` in `pimd.conf` (or even no `pimd.conf`) -- no interfaces are detected and `pimd` segfaults :-1: 

Also, the `goto init_vif_list` is unbounded and makes it possible for `pimd` to get stuck in an endless loop :-1: 

I'll try to fix these myself, and take it as a lesson to myself to actually test future pull requests, but if I cannot find a satisfactory fix I will revert this change.

**Comment by troglobit** [comment]
I've made the following changes:

dcdfd65 : Must use `-r` command line option to retry forever in `-N` operation  
e9d19f0 : Allow empty `pimd.conf` and regular non`-N` operation, do not skip `!IFF_UP` interfaces

If you find any regressions to your particular use-case, please file a separate issue first so we can discuss a fix.

**Pull request merged**